common/SmokeTests/SmokeTest/Program.cs

```diff
@@ -1,37 +1,214 @@
+// its functionality.
+//
+// Execution will be logged to the console, using standard output for informational messages
+// and standard error for errors.   A return code of "0" will indicate a successful run; a
+// non-zero return indicates a test failure.
+
+const string AssemblyFileMask = "*.dll";
+
+LogInformation($"Smoke Test run starting...{ Environment.NewLine }");
+
+var returnCode = 0;
+
+// Starting with the entry assembly as the root, walk all references, both
+// direct and transient to process them.
+
+foreach (var assembly in LoadAssemblies(Assembly.GetEntryAssembly(), AssemblyFileMask))
 {
-    class Program
+    LogInformation($"Processing assembly: [{ assembly.FullName }]...");
+
+    var typeCount = 0;
+
+    try
     {
-        static async Task<int> Main(string[] args)
+        // Process each type in the assembly.  The call to "ProcessType" is safe and
+        // will not throw.
+
+        foreach (var type in assembly.GetTypes())
         {
-            Console.WriteLine("SMOKE TEST FOR TRACK 2 LIBRARIES");
-
-            await KeyVaultTest.RunTests();
-            await BlobStorageTest.RunTests();
-            await EventHubsTest.RunTests();
-            // Temporarily disable Cosmos DB smoke tests
-            // Issue to re-enable:
-            // https://github.com/Azure/azure-sdk-for-net/issues/11297
-            // await CosmosDBTest.RunTests();
-
-            Console.WriteLine();
-            Console.WriteLine("LIBRARY TESTS COMPLETE");
-            Console.WriteLine();
-            Console.WriteLine("SMOKE TEST FOR TRACK 2 SAMPLES");
-
-            await IotHubConnectionTests.RunTests();
-
-            Console.WriteLine("SAMPLES TESTS COMPLETE");
-            Console.WriteLine();
-            return 0;
+            ++typeCount;
+            ProcessType(type);
         }
     }
+    catch (Exception ex)
+    {
+        ReportError();
+        LogError($"{Environment.NewLine }[{ assembly.FullName }] could not be fully processed.  Error: { ex }{ Environment.NewLine }");
+    }
+
+    LogInformation($"Completed assembly: [{ assembly.FullName }].  { typeCount } types inspected.{ Environment.NewLine }");
+}
+
+// End the test run and report on the state.
+
+LogInformation($"{ Environment.NewLine }Smoke Test run complete.");
+
+if (returnCode == 0)
+{
+    LogInformation("Passed.", ConsoleColor.Green);
 }
+else
+{
+    LogInformation("Failed.", ConsoleColor.Red);
+}
+
+return returnCode;
+
+// Local function definitions
+
+void ReportError() => returnCode = -1;
+void LogError(string message, ConsoleColor? color = default) => Console.Error.WriteLine(message);
+
+void LogInformation(string message, ConsoleColor? color = default)
+{
+    Console.ForegroundColor = color ?? Console.ForegroundColor;
+    Console.WriteLine(message);
+    Console.ResetColor();
+}
+
+void ProcessType(Type type)
+{
+    // Inspect the members of the type, walking each to force metadata to be loaded.  If a parameterless
```

Comment:
Are you trying to load all types in the assemblies? I thought we were planning to load the client types only.  As a side question do the client types end up having the default constructors so we can load them?

Reply:
As I understand it, the act of querying the metadata for types via reflection is technically enough to force the transient dependencies to be resolved and would sniff out issues.  I'm creating instances just to go one step further and ensure that we instantiate at least one type from each reference.   

The clients themselves do not have public parameterless constructors - but we know that the options for each client do, so we'll get at least one type per Azure library reference that gets created.

Reply:
With respect to loading the client types, there's no general way to do so.  We'd have to have some deeper insight to identify clients via naming patterns, and then be able to reverse-engineer the parameters via metadata and somehow provide arguments not only of the correct type, but also values that would pass validation.   We'd really need to have a map with a logic specific to each client type that could tell the smoke tests how to construct it appropriately - and one of our goals was to avoid having to code and maintain specific logic like that.

When Alex and I talked this over, we landed on loading up all direct and transient assembly references and then walking the types for each via reflection and inspecting members.  Because it was a simple step from there to "activate non-generic parameterless things", I did so as well.

Reply:
At some point, I remembered that our guidelines call for client types to support a protected parameterless constructor for mocking.   I took advantage of this to instantiate clients that are compliant with this.